File: sdk/python/test_restrictions_client.py
from __future__ import annotations

import pytest
from typing import Optional

from lichen import (
    AccountAssetRestrictionTarget,
    Keypair,
    RestrictAccountParams,
    RestrictionGovernanceClient,
    ResumeBridgeRouteParams,
    SetFrozenAssetAmountParams,
    TransferRestrictionParams,
)


class FakeConnection:
    def __init__(self) -> None:
        self.calls: list[tuple[str, list[object]]] = []

    async def rpc_request(self, method: str, params: Optional[list[object]] = None):
        payload = params or []
        self.calls.append((method, payload))
        return {
            "method": method,
            "params": payload,
            "unsigned": method.startswith("build"),
        }


@pytest.mark.asyncio
async def test_restriction_client_read_and_preflight_payloads() -> None:
    connection = FakeConnection()
    client = RestrictionGovernanceClient(connection)
    account = Keypair.from_seed(bytes(range(32))).pubkey()
    recipient = Keypair.from_seed(bytes(range(1, 33))).pubkey()
    asset = Keypair.from_seed(bytes(range(2, 34))).pubkey()

    await client.get_account_restriction_status(account)
    await client.get_restriction_status(AccountAssetRestrictionTarget(account=account, asset=asset))
    await client.list_restrictions(limit=10, after_id=2)
    await client.can_transfer(
        TransferRestrictionParams(
            from_=account,
            to=recipient,
            asset="native",
            amount=123,
        )
    )

    assert connection.calls == [
        ("getAccountRestrictionStatus", [str(account)]),
        (
            "getRestrictionStatus",
            [
                {
                    "type": "account_asset",
                    "account": str(account),
                    "asset": str(asset),
                }
            ],
        ),
        ("listRestrictions", [{"limit": 10, "after_id": 2}]),
        (
            "canTransfer",
            [
                {
                    "from": str(account),
                    "to": str(recipient),
                    "asset": "native",
                    "amount": 123,
                }
            ],
        ),
    ]


@pytest.mark.asyncio
async def test_restriction_client_builder_payloads() -> None:
    connection = FakeConnection()
    client = RestrictionGovernanceClient(connection)
    proposer = Keypair.from_seed(bytes(range(10, 42))).pubkey()
    authority = Keypair.from_seed(bytes(range(11, 43))).pubkey()
    account = Keypair.from_seed(bytes(range(12, 44))).pubkey()
    asset = Keypair.from_seed(bytes(range(13, 45))).pubkey()

    restrict = await client.build_restrict_account_tx(
        RestrictAccountParams(
            proposer=proposer,
            governance_authority=authority,
            account=account,
            mode="outgoing_only",
            reason="testnet_drill",
            recent_blockhash="abc123",
            expires_at_slot=500,
        )
    )
    frozen = await client.build_set_frozen_asset_amount_tx(
        SetFrozenAssetAmountParams(
            proposer=proposer,
            governance_authority=authority,
            account=account,
            asset=asset,
            amount=100,
            reason="stolen_funds",
        )
    )
    resumed = await client.build_resume_bridge_route_tx(
        ResumeBridgeRouteParams(
            proposer=proposer,
            governance_authority=authority,
            chain="ethereum",
            asset="eth",
            restriction_id=5,
            lift_reason="testnet_drill_complete",
        )
    )

    assert restrict["unsigned"] is True
    assert frozen["unsigned"] is True
    assert resumed["unsigned"] is True
    assert connection.calls == [
        (
            "buildRestrictAccountTx",
            [
                {
                    "proposer": str(proposer),
                    "governance_authority": str(authority),
                    "recent_blockhash": "abc123",
                    "reason": "testnet_drill",
                    "expires_at_slot": 500,
                    "account": str(account),
                    "mode": "outgoing_only",
                }
            ],
        ),
        (
            "buildSetFrozenAssetAmountTx",
            [
                {
                    "proposer": str(proposer),
                    "governance_authority": str(authority),
                    "reason": "stolen_funds",
                    "account": str(account),
                    "asset": str(asset),
                    "amount": 100,
                }
            ],
        ),
        (
            "buildResumeBridgeRouteTx",
            [
                {
                    "proposer": str(proposer),
                    "governance_authority": str(authority),
                    "chain": "ethereum",
                    "asset": "eth",
                    "restriction_id": 5,
                    "lift_reason": "testnet_drill_complete",
                }
            ],
        ),
    ]


@pytest.mark.asyncio
async def test_restriction_client_rejects_invalid_u64_values() -> None:
    connection = FakeConnection()
    client = RestrictionGovernanceClient(connection)
    account = Keypair.from_seed(bytes(range(20, 52))).pubkey()

    with pytest.raises(ValueError, match="restriction_id must be a u64-safe integer value"):
        await client.get_restriction(-1)

    with pytest.raises(ValueError, match="amount must be a u64-safe integer value"):
        await client.can_transfer(
            {
                "from": account,
                "to": account,
                "asset": "native",
                "amount": True,
            }
        )

    assert connection.calls == []

File: sdk/js/test_restrictions_client.js
import assert from 'node:assert/strict';
import {
  BRIDGE_ASSETS,
  BRIDGE_CHAINS,
  Connection,
  PublicKey,
  RestrictionGovernanceClient,
} from './dist/index.js';

class MockConnection {
  calls = [];

  async rpcRequest(method, params = []) {
    this.calls.push({ method, params });
    return {
      method,
      params,
      unsigned: true,
      encoding: 'base64',
      wire_format: 'lichen_tx_v1',
      tx_type: 'native',
      transaction_base64: 'AA==',
      transaction: 'AA==',
      wire_size: 1,
      message_hash: '00',
      signature_count: 0,
      recent_blockhash: '00',
      proposer: '',
      governance_authority: '',
      action_label: 'restrict',
      action: {},
      instruction: {
        program_id: '',
        accounts: [],
        instruction_type: 34,
        governance_action_type: 10,
        data_hex: '',
      },
    };
  }
}

function key(byte) {
  return new PublicKey(new Uint8Array(32).fill(byte));
}

const connection = new MockConnection();
const client = new RestrictionGovernanceClient(connection);
const proposer = key(1);
const authority = key(2);
const account = key(3);
const recipient = key(4);
const asset = key(5);

await client.getAccountRestrictionStatus(account);
assert.deepEqual(connection.calls.at(-1), {
  method: 'getAccountRestrictionStatus',
  params: [account.toBase58()],
});

await client.getRestrictionStatus({ type: 'account_asset', account, asset: 'native' });
assert.deepEqual(connection.calls.at(-1), {
  method: 'getRestrictionStatus',
  params: [{ type: 'account_asset', account: account.toBase58(), asset: 'native' }],
});

await client.canTransfer({ from: account, to: recipient, asset, amount: 25n });
assert.deepEqual(connection.calls.at(-1), {
  method: 'canTransfer',
  params: [{
    from: account.toBase58(),
    to: recipient.toBase58(),
    asset: asset.toBase58(),
    amount: '25',
  }],
});

await client.buildRestrictAccountTx({
  proposer,
  governanceAuthority: authority,
  account,
  mode: 'outgoing_only',
  reason: 'testnet_drill',
  evidenceHash: 'aa'.repeat(32),
  expiresAtSlot: 123n,
  recentBlockhash: 'bb'.repeat(32),
});
assert.deepEqual(connection.calls.at(-1), {
  method: 'buildRestrictAccountTx',
  params: [{
    proposer: proposer.toBase58(),
    governance_authority: authority.toBase58(),
    recent_blockhash: 'bb'.repeat(32),
    reason: 'testnet_drill',
    evidence_hash: 'aa'.repeat(32),
    expires_at_slot: '123',
    account: account.toBase58(),
    mode: 'outgoing_only',
  }],
});

await client.buildSetFrozenAssetAmountTx({
  proposer,
  governanceAuthority: authority,
  account,
  asset,
  amount: 500n,
  reason: 'testnet_drill',
});
assert.deepEqual(connection.calls.at(-1), {
  method: 'buildSetFrozenAssetAmountTx',
  params: [{
    proposer: proposer.toBase58(),
    governance_authority: authority.toBase58(),
    reason: 'testnet_drill',
    account: account.toBase58(),
    asset: asset.toBase58(),
    amount: '500',
  }],
});

await client.buildResumeBridgeRouteTx({
  proposer,
  governanceAuthority: authority,
  chain: BRIDGE_CHAINS.NEOX,
  asset: BRIDGE_ASSETS.GAS,
  restrictionId: 12,
  liftReason: 'testnet_drill_complete',
});
assert.deepEqual(connection.calls.at(-1), {
  method: 'buildResumeBridgeRouteTx',
  params: [{
    proposer: proposer.toBase58(),
    governance_authority: authority.toBase58(),
    chain: 'neox',
    asset: 'gas',
    restriction_id: 12,
    lift_reason: 'testnet_drill_complete',
  }],
});

await assert.rejects(
  () => client.buildLiftRestrictionTx({
    proposer,
    governanceAuthority: authority,
    restrictionId: -1,
    liftReason: 'false_positive',
  }),
  /restrictionId must be a u64-safe integer value/,
);

const originalFetch = globalThis.fetch;
const rpcRequests = [];
globalThis.fetch = async (_url, init) => {
  const body = JSON.parse(init.body);
  rpcRequests.push(body);
  const page = rpcRequests.length === 1
    ? {
        contracts: [{ program_id: 'first', code_size: 1, is_executable: true, has_abi: false, abi_functions: 0, code_hash: '', version: 1, lifecycle_status: 'active', lifecycle_updated_slot: 0, lifecycle_effective_at_slot: 0 }],
        count: 1,
        has_more: true,
        next_cursor: 'cursor-1',
      }
    : {
        contracts: [{ program_id: 'second', code_size: 1, is_executable: true, has_abi: false, abi_functions: 0, code_hash: '', version: 1, lifecycle_status: 'active', lifecycle_updated_slot: 0, lifecycle_effective_at_slot: 0 }],
        count: 1,
        has_more: false,
        next_cursor: null,
      };
  return new Response(JSON.stringify({ jsonrpc: '2.0', id: body.id, result: page }), {
    status: 200,
    headers: { 'Content-Type': 'application/json' },
  });
};

try {
  const sdkConnection = new Connection('http://mock-rpc');
  const contracts = await sdkConnection.getAllContractsAll(1);
  assert.deepEqual(contracts.map((entry) => entry.program_id), ['first', 'second']);
  assert.deepEqual(rpcRequests.map((request) => request.params), [
    [{ limit: 1 }],
    [{ limit: 1, cursor: 'cursor-1' }],
  ]);
} finally {
  globalThis.fetch = originalFetch;
}
